threadloop/threadloop.py

```diff
@@ -124,7 +125,9 @@ def on_done(tornado_future):
             future.set_exception(exception)
 
         self._io_loop.add_callback(
-            lambda: fn(*args, **kwargs).add_done_callback(on_done)
+            lambda: gen.maybe_future(
+                fn(*args, **kwargs)
```

Comment:
That's a drawback of `maybe_future`. It operates on the value returned by the function, so if the function itself fails to return a value, you're going to throw an exception. This is usually not a problem if you're also doing a `yield` because you expect the exception -- either from the future or from the call, but if you're doing anything more complex you have to handle the other cases.

    yield maybe_future(f(...))

Can fail in either calling `f` or in the `future.result()` called by `yield`. You're fine either way because you expect the exception there.

    maybe_future(f(...)).add_done_callback(..)

If `f` fails, you're not adding any callback. Your callback will never be called.